### ISSUE TITLE
`no-member` false positive when replacing dict values with objects of different types

### ISSUE THREAD

**Issue by gunit367** [PyCQA/pylint#4864]
### Bug description

Code runs successfully, but incorrectly lints with a no-member failure.

```python

class Class1():
    def __init__(self):
        d = {
            1 : self.func1,
            2 : self.func2,
        }
        for num in d:
            newObj = Class2()
            d[num] = newObj
        newObj.func3()

    def func1(self):
        print("func1")
    def func2(self):
        print("func2")

class Class2():
    def __init__(self):
        pass
    def func3(self):
        print("func3")

def main():
    o = Class1()
    print("Made object successfully")

if __name__ == '__main__':
    main()
```

### Command used

```shell
pylint -E tmp.py
```

### Pylint output

```shell
******************* Module tmp
E: 15, 8: Method 'func1' has no 'func3' member (no-member)
```

### Expected behavior

No errors should have occurred.

### Pylint version

```shell
pylint 1.8.3
astroid 1.6.0
python 3.6.9
```

### OS / Environment

Ubuntu 18.04.5 LTS

**Comment by Pierre-Sassoulas** [comment]
Pylint 1.8.3 is very old, could you try with pylint 2.9.6 ?

**Comment by Pierre-Sassoulas** [comment]
This does not happen with the latest pylint, and we don't maintain old version of pylint.
